keps/sig-network/20190424-NodeLocalDNS-beta-proposal.md

```diff
@@ -0,0 +1,173 @@
+---
+title: graduate-nodelocaldns-to-beta
+authors:
+  - "@prameshj"
+owning-sig: sig-network
+reviewers:
+  - "@bowei"
+  - "@thockin"
+  - "@johnbelamaric"
+approvers:
+  - "@bowei"
+  - "@thockin"
+creation-date: 2019-04-24
+last-updated: 2019-04-25
+status: implementable
+see-also:
+  - "/keps/sig-network/0030-nodelocal-dns-cache.md"
+---
+
+# Graduate NodeLocal DNSCache to beta
+
+## Table of Contents
+
+A table of contents is helpful for quickly jumping to sections of a KEP and for highlighting any additional information provided beyond the standard KEP template.
+[Tools for generating][] a table of contents from markdown are available.
+
+- [Title](#title)
+  - [Table of Contents](#table-of-contents)
+  - [Summary](#summary)
+  - [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+  - [Proposal](#proposal)
+    - [Risks and Mitigations](#risks-and-mitigations)
+  - [Design Details](#design-details)
+    - [Test Plan](#test-plan)
+    - [Graduation Criteria](#graduation-criteria)
+  - [Implementation History](#implementation-history)
+  - [Alternatives [optional]](#alternatives)
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+
+[NodeLocal DNSCache](https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/dns/nodelocaldns/README.md) is an addon that runs a dnsCache pod as a daemonset to improve clusterDNS performance and reliability. The feature has been in Alpha since 1.13 release. This document lays out the plan to promote it to beta.
+
+## Motivation
+
+NodeLocal DNSCache has been in Alpha for the past 2 releases and users have deployed it to fix DNS performance issues.
+- [Example 1](https://github.com/kubernetes/kubernetes/issues/56903#issuecomment-485353223)
+- [Example 2](https://github.com/kubernetes/kubernetes/issues/45363#issuecomment-443019910)
+
+Based on the feedback so far, we feel the feature is ready to be graduated to beta.
+
+### Goals
+
+Graduate NodeLocal DNSCache to beta.
+
+## Proposal
+
+N.B. Although CoreDNS is now the default DNS server on Kubernetes clusters, this document still uses the name kube-dns since the service name is still the same.
+
+Based on the initial feedback for NodeLocal DNSCache feature, HA seems to be the common ask. 
```

Comment:
I'm skeptical.  I 100% agree that it is a little different than kubelet or kube-proxy (this is "data plane" vs "control plane").  That said, this proposal is kind of going to heroic lengths to get within a single failure-domain.   We have a number of node-agents and they are all, more or less, subject to this problem.

What you propose here only reduces the outage window, and carries its own risks.  We know that running iptables can consume a lot of memory (yay iptables), which means we have to set aside a significant amount of RAM for this, or it will actually make the system less stable (OOM).

This also removes our ability to do anything "clever" in the cache, such as the autopath proposal.  It requires that the pod <-> cache protocol be the same as cache <-> upstream.  This makes me sad, because I like the autopath proposal.

Additionally this doesn't work for IPVS, which is a major problem, and it sounds like we don't have a good answer.

Given all this, can we instead do something like SO_REUSEPORT ?  If we ensure the cache is SO_REUSEPORT enabled, users who want HA can run 2 copies of the cache (HA, for the low, low price of 2x?), and users who don't care can just ... not ?

Reply:
Do we need HA for Beta?  I propose we push it beyond beta.  It's way more valuable to get this into more hands ASAP, IMO.

Reply:
Put another way - HA capability sounds great, but we should not block Beta on it.

Reply:
Not a hard requirement, I put it as a criterion since it was the most asked question about the feature. I am ok with decoupling this with the Beta Graduation criteria and continue working on a solution.

Reply:
> Do we need HA for Beta? 

Does beta mean this feature is enabled by default in new clusters?
Can this be "backported" or will be available only in newer versions?

Reply:
feature will not be enabled by default on new clusters. It can be backported to older versions.

Reply:
As it is not a bugfix, it probably wouldn't make sense to backport. However, given it has almost no dependencies, we can give users a very easy way to configure their cluster to use it.